homeassistant/components/squeezebox/config_flow.py

```diff
@@ -0,0 +1,230 @@
+"""Config flow for Logitech Squeezebox integration."""
+import asyncio
+import logging
+
+from pysqueezebox import Server, async_discover
+import voluptuous as vol
+
+from homeassistant import config_entries, exceptions
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_USERNAME,
+    HTTP_UNAUTHORIZED,
+)
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+# pylint: disable=unused-import
+from .const import DEFAULT_PORT, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): str,
+        vol.Required(CONF_PORT, default=DEFAULT_PORT): int,
+        vol.Optional(CONF_USERNAME): str,
+        vol.Optional(CONF_PASSWORD): str,
+    }
+)
+
+TIMEOUT = 5
+
+
+class SqueezeboxConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Logitech Squeezebox."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize an instance of the squeezebox config flow."""
+        self.data_schema = DATA_SCHEMA
+        self.discovery_info = None
+
+    async def _discover(self, uuid=None):
+        """
+        Discover an unconfigured LMS server.
+
+        Parameters:
+            uuid: search for this uuid (optional)
+        """
+        self.discovery_info = None
+        discovery_event = asyncio.Event()
+
+        def _discovery_callback(server):
+            if server.uuid:
+                if uuid:
+                    # ignore non-matching uuid
+                    if server.uuid != uuid:
+                        return
+                else:
+                    # ignore already configured uuids
+                    for entry in self._async_current_entries():
+                        if entry.unique_id == server.uuid:
+                            return
+                self.discovery_info = {
+                    CONF_HOST: server.host,
+                    CONF_PORT: server.port,
+                    "uuid": server.uuid,
+                }
+                _LOGGER.debug("Discovered server: %s", self.discovery_info)
+                discovery_event.set()
+
+        discovery_task = self.hass.async_create_task(
+            async_discover(_discovery_callback)
+        )
+
+        await discovery_event.wait()
+        discovery_task.cancel()  # stop searching as soon as we find server
+
+        # update with suggested values from discovery
+        self.data_schema = vol.Schema(
```

Comment:
There is a lot of similar code to build the schema.  You can pass discovery_info/user_input to a function to build it in pieces to avoid repeating yourself.

Example: https://github.com/home-assistant/core/pull/36827/commits/fc352e93b020547a6afd8459ac65ff4fdd0bdaab

Reply:
Ok, implemented a `base_schema` function that builds the schema based on `discovery_info` or `user_input`.

Reply:
It looks like there is still quite a bit of schema duplication